src/gluonts/transform.py

```diff
@@ -1242,6 +1280,169 @@ def flatmap_transform(
+    target_field
+        field containing the target
+    start_field
+        field containing the start date of the of the point process observation
+    end_field
+        field containing the end date of the point process observation
+    forecast_start_field
+        output field that will contain the time point where the forecast starts
+    """
+
+    def __init__(
+        self,
+        past_interval_length: float,
+        future_interval_length: float,
+        train_sampler: ContinuousTimePointSampler,
+        target_field: str = "target",
+        start_field: str = "start",
+        end_field: str = "end",
+        forecast_start_field: str = "forecast_start",
+    ) -> None:
+
+        assert (
+            future_interval_length > 0
+        ), "Prediction interval must have length greater than 0."
+
+        self.train_sampler = train_sampler
+        self.past_interval_length = past_interval_length
+        self.future_interval_length = future_interval_length
+        self.target_field = target_field
+        self.start_field = start_field
+        self.end_field = end_field
+        self.forecast_start_field = forecast_start_field
+
+    # noinspection PyMethodMayBeStatic
+    def _mask_sorted(self, a: np.ndarray, lb: float, ub: float):
+        start = np.searchsorted(a, lb)
+        end = np.searchsorted(a, ub)
+        return np.arange(start, end)
+
+    def flatmap_transform(
+        self, data: DataEntry, is_train: bool
+    ) -> Iterator[DataEntry]:
+
+        assert data[self.start_field].freq == data[self.end_field].freq
+
+        total_interval_length = (
+            data[self.end_field] - data[self.start_field]
+        ) / data[self.start_field].freq.delta
+
+        # sample forecast start times in continuous time
+        if is_train:
+            if total_interval_length < (
+                self.future_interval_length + self.past_interval_length
+            ):
+                sampling_times: np.ndarray = np.array([])
+            else:
+                sampling_times = self.train_sampler(
+                    self.past_interval_length,
+                    total_interval_length - self.future_interval_length,
+                )
+        else:
+            sampling_times = np.array([total_interval_length])
+
+        # keep a copy of the target
+        d = data.copy()
+
+        ia_times = d[self.target_field][0, :]
+        marks = d[self.target_field][1:, :]
+
+        ts = np.cumsum(ia_times)
+        assert ts[-1] < total_interval_length, (
+            "Target interarrival times provided are inconsistent with "
+            "start and end timestamps."
+        )
+
+        # select field names that will be included in outputs
+        keep_cols = {
+            k: v
+            for k, v in d.items()
+            if k not in [self.target_field, self.start_field, self.end_field]
+        }
+
+        for future_start in sampling_times:
+
+            r = dict()
```

Comment:
`d` was initialized outside of the loop here (this may be a spurious copy, now that I look at it). `r`is the output `dict`. 

I'm not a fan of the style in `InstanceSplitter` where d is "copied" once per each output. However this is (probably?) a shallow copy. Moreover, features an explicit 
```
del d[ts_field]
```

I thought this style was a little safer, or maybe I'm missing something here?

In any case, ditto for `DataEntry` :+1:

Reply:
The only reason I would regard this as safer is if you have to do multiple transformations that rely on one original entry on in the `DataEntry` (which would make sense to have it immutable in that case). I'm not a fan because it introduces more code than necessary, but tastes differ :-).

Reply:
Yup. I agree. I thought in `flatmap_transformation` this is indeed the intention that multiple entries derive from a single input. Hence the extra code, to stress that the input is immutable.

Just crossing the t's here to make sure :)